tests/MongoDB.Bson.Tests/Specifications/uuid/prose-tests/ExplicitDecodingTests.cs

```diff
@@ -0,0 +1,117 @@
+                exception.Should().BeOfType<ArgumentException>();
+
+                var result = binaryData.ToGuid(GuidRepresentation.CSharpLegacy);
+                result.Should().Be(new Guid("00112233445566778899aabbccddeeff"));
+            }
+        }
+
+        [Fact]
+        public void BsonBinaryData_with_sub_type_3_and_Java_byte_order_tests()
+        {
+            using (TemporaryGuidRepresentationMode.V3())
+            {
+                var bytes = BsonUtils.ParseHexString("7766554433221100ffeeddccbbaa9988");
+                var binaryData = new BsonBinaryData(bytes, BsonBinarySubType.UuidLegacy);
+
+                var exception = Record.Exception(() => binaryData.ToGuid());
+                exception.Should().BeOfType<InvalidOperationException>();
+
+                exception = Record.Exception(() => binaryData.ToGuid(GuidRepresentation.Standard));
+                exception.Should().BeOfType<InvalidOperationException>();
+
+                exception = Record.Exception(() => binaryData.ToGuid(GuidRepresentation.Unspecified));
+                exception.Should().BeOfType<ArgumentException>();
+
+                var result = binaryData.ToGuid(GuidRepresentation.JavaLegacy);
+                result.Should().Be(new Guid("00112233445566778899aabbccddeeff"));
+            }
+        }
+
+        [Fact]
+        public void BsonBinaryData_with_sub_type_3_and_Python_byte_order_tests()
+        {
+            using (TemporaryGuidRepresentationMode.V3())
+            {
+                var bytes = BsonUtils.ParseHexString("00112233445566778899aabbccddeeff");
+                var binaryData = new BsonBinaryData(bytes, BsonBinarySubType.UuidLegacy);
+
+                var exception = Record.Exception(() => binaryData.ToGuid());
+                exception.Should().BeOfType<InvalidOperationException>();
+
+                exception = Record.Exception(() => binaryData.ToGuid(GuidRepresentation.Standard));
+                exception.Should().BeOfType<InvalidOperationException>();
+
+                exception = Record.Exception(() => binaryData.ToGuid(GuidRepresentation.Unspecified));
+                exception.Should().BeOfType<ArgumentException>();
+
+                var result = binaryData.ToGuid(GuidRepresentation.PythonLegacy);
+                result.Should().Be(new Guid("00112233445566778899aabbccddeeff"));
+            }
+        }
+
+        [Fact]
+        public void BsonBinaryData_with_sub_type_4_tests()
+        {
+            using (TemporaryGuidRepresentationMode.V3())
+            {
+                var guid = new Guid("00112233445566778899aabbccddeeff");
+                var bytes = GuidConverter.ToBytes(guid, GuidRepresentation.Standard);
+                var binaryData = new BsonBinaryData(bytes, BsonBinarySubType.UuidStandard);
+
+                var exception = Record.Exception(() => binaryData.ToGuid(GuidRepresentation.Unspecified));
```

Comment:
Not sure that I fully understand you here..
Let's say the exception in line 100 is correct, because `GuidRepresentation.Unspecified` is invalid for this method for any condition, but what is different in this method: https://github.com/rstam/mongo-csharp-driver/blob/csharp2724b/src/MongoDB.Bson/ObjectModel/GuidConverter.cs#L137
As far as I can see, this argument is invalid for all conditions as well..

In one case we throw `ArgumentException` in another - `InvalidOperationException`.

Reply:
Line 100 throws an ArgumentException because Unspecified is ALWAYS an invalid argument to this overload of ToGuid.

Line 105 throws an InvalidOperationException because while the call to ToGuid is valid with the given guidRepresentation in some cases it is not valid given the current state of the BsonBinaryData object (because subtype is 4).

Line 137 in GuidConverter.cs (a different file from this one) is wrong, but I don't know think that we should change it now, at least not as part of this ticket.